### PR TITLE
Fixes for LEDC driven PWM channels on ESP32

### PR DESCRIPTION
**Pull request by pkendall64**
As part of the PWM (ESP32-S3) refactor I broke the LEDC PWM support for regular ESP32 devices. This only affected channels > 12 as the first 12 are driven by the MCPWM module.

~Also, as a part of this I output the failsafe values at startup rather than no-signal as the ESP32 boot ROM outputs PWM signal on some pins and causes some servos to go mad and push into areas past the limits of the model they are connected to.~

EDIT: I removed the part where I set the failsafe positions on startup as this should really be a separate PR after discussion about the best way to achieve the desired outcome.

### CODE REVIEW

**Comment by CapnBry** [comment]
This change ignores the fact that ESCs need no signal on startup so that they can be calibrated after the RX connects. The couple I tried seem to ignore the startup UART stuff enough for me to do ESC calibration. This isn't a bugfix it is a design change but I suppose we have to to work around the terrible design decision by espressif to spam the UART? I'd love to hear other options though.

**Comment by pkendall64** [comment]
I agree with you that this is not the best way to do this. BTW, it's not just the default serial TX pin either!
On the ESP32; GPIO0, 5, 14, 15 also spam out a PWM signal on bootup from the bootloader ROM!

Perhaps the best solution is to allow the user to select what to do for each pin on initialisation i.e. "No signal" or "Failsafe".

**Comment by CapnBry** [comment]
Yeah I can see it both ways so it is tough to figure out what is the right thing to do. However, if servos are losing their minds due to the ROM spam on startup and this resolves that then I think you should leave the set failsafe in.

An ESC can still be configured by leaving it unplugged until connection as a workaround, but then that means if your ESC is powering the RX then you'll need another power source in a sort of chicken and egg situation. That's a one time thing though so not a huge deal. Does setting the failsafe position resolve the issue entirely, or does it still go to full deflection and break stuff, then go where it was supposed to a second later trying to pretend like nothing happened?

I have some work that needs doing to unify the failsafe procedure and positions across all serialio types, which is needed so teamrace can be merged, so I'll just put the "use the configured failsafe mode on startup" idea in that todo list.

**Comment by schugabe** [comment]
> An ESC can still be configured by leaving it unplugged until connection as a workaround, but then that means if your ESC is powering the RX then you'll need another power source in a sort of chicken and egg situation. That's a one time thing though so not a huge deal. Does setting the failsafe position resolve the issue entirely, or does it still go to full deflection and break stuff, then go where it was supposed to a second later trying to pretend like nothing happened?

Are you talking about the calibration of ESCs? That could be solved by setting the FS value to a high value to get the ESC in the config mode and afte that you have the connection to move the stick to the correct positions.

I think I would never expect the rx to output random stuff. So sending FS values would be my expected behaviour.

**Review by schugabe — APPROVED** [review]

**Comment by CapnBry** [comment]
If I want to calibrate an ESC, I want to plug it in with the throttle high, wait for beeps, and lower the throttle. I don't want to dig into the Lua, other devices, set failsafe position, unplug, replug, wait for beeps, lower throttle, go back into the lua, other devices, set failsafe position. That's an insanely complicated procedure that has me digging through menus and if I don't remember to lower it again (or it didn't go through because there's really not any sort of guarantee it did) now my failsafe is full throttle. This can be especially problematic if the telemetry bit is out of sync due to rebooting the receiver, which might cause it to take some time to sync up, or if the user stayed in the Lua, it could just ignore changing the failsafe back. That's a solution, but I feel it is slightly more dangerous due to these factors.

And the RX still does output random stuff, but with this change it outputs random stuff, then goes to another position faster than it does currently. It still does the bad thing, it just covers it up by doing something different soon after. 😅 Did my servo just go full deflection and jam up then go back?! Nope. Nothing to see here!

**Comment by CapnBry** [comment]
Can you fix this too?
```
diff --git a/src/lib/PWM/PWM_ESP32.cpp b/src/lib/PWM/PWM_ESP32.cpp
index b9521f6a..17f7cf0d 100644
--- a/src/lib/PWM/PWM_ESP32.cpp
+++ b/src/lib/PWM/PWM_ESP32.cpp
@@ -190,7 +190,7 @@ void PWMController::release(pwm_channel_t channel)
         mcpwm_stop(mcpwm_config[ch].unit, mcpwm_config[ch].timer);
         mcpwm_frequencies[ch] = 0;
     }
-    else if (IS_MCPWM_CHANNEL(channel))
+    else if (IS_LEDC_CHANNEL(channel))
     {
         auto ch = LEDC_CHANNEL(channel);
         ledcDetachPin(ledc_config[ch].pin);
```

File: src/lib/PWM/PWM_ESP32.cpp [hunk 1]
```diff
@@ -128,7 +137,10 @@ pwm_channel_t PWMController::allocate(uint8_t pin, uint32_t frequency)
             .duty_mode = MCPWM_DUTY_MODE_0,
             .counter_mode = MCPWM_UP_COUNTER,
         };
-        mcpwm_gpio_init(mcpwm_config[channel].unit, mcpwm_config[channel].signal, pin);
+        if (mcpwm_gpio_init(mcpwm_config[channel].unit, mcpwm_config[channel].signal, pin) != ESP_OK)
+        {
+            DBGLN("failed %d", pin);
```
Comment (CapnBry) [inline src/lib/PWM/PWM_ESP32.cpp:142]: Maybe add a tiny bit more info here? Seeing just `failed 7` in a log might be hard to track down. Throw a `mcpwm_gpio_init` in front?

**Review by CapnBry — CHANGES_REQUESTED** [review]
I don't have any S3 hardware to test with but this plus the changes I've listed in the discussion.

I'd add back in the failsafe on startup, I don't particularly like it but I don't think there's more to discuss. Let's hide that the hardware does something bad by jumping right to failsafe. I only worry that the default failsafe values might set someone's throttle high, but if it is on boot then it should just go into programming mode right? :-D

**Comment by pkendall64** [comment]
@CapnBry I fixed the issues you pointed out and I found another one where the array initialisation was wrong. Perhaps that was causing your problem?

**Comment by pkendall64** [comment]
And I reinstated the set failsafe as well.

**Comment by CapnBry** [comment]
I know most of these aren't in this PR directly, but since this is about the PWM unit

### PWM_ESP32.cpp
* ledc_config initializes first item to {-1,0,0}, which is never used and the other items are still {0,0,0}. This moves the whole array into flash from bss and isn't needed at all. Recommend removing initialization.
* ledcTimerConfigs not declared static
* Can save 692 bytes by replacing log2f with (and remove math.h)
```
            uint8_t bits = 0;
            uint32_t clock = 80000000U / frequency; // APB clk src is 80Mhz
            while (clock >>= 1)
                ++bits;
```

### devServoOutput.cpp
* SERVO_PINS / PWM_CHANNELS should be normal camel case variable names as they are not macros / constants (this was my bad)
* Remove -1 initializer from SERVO_PINS/PWM_CHANNELS declaration to move it to bss and initialize it in init only
* Don't use two loops to init SERVO_PINS/PWM_CHANNELS. First, SERVO_PINS is filled by the following loop, and the constant it uses is UNDEF_PIN, but setting it twice isn't necessary. Just move `PWM_CHANNELS[0] = -1` to the first line of the other loop and remove this second loop.
* Remove unused dshotProtocol variable. If this is intended to be a constant make it a constant instead
* rmtCH / RMT_MAX_CHANNELS variables do not need to be global, move them to where they are use. RMT_MAX_CHANNELS is also a constant but declared as a variable
* Questionable: I'm not sure about all the debug logging in the initialize, init is to initialize variables to a zero state and set pin modes but in this instance it is doing a lot more. Probably ok, but it seems weird to see logging before everything else reporting all the setup it did.

My main issue though is how my True Diversity PWM receiver is having so many problems and losing connection, see our dev thread "...Receiver So Broke". I feel like that needs to be resolved as a blocker but I have no clue what is causing it. Alignment of some variable issue? I've spent a whole day trying to find it but haven't come up with anything and I am out of ideas.

**Review by CapnBry — APPROVED** [review]
Looks good! Tested on SuperP and then again forcing it to use LEDC first and servos wiggled properly. Did not put them on the oscilloscope but mechanically behaved correctly.

**Pull request merged**